scipy/optimize/_direct.py

```diff
@@ -0,0 +1,203 @@
+)
+
+import numpy as np
+from scipy.optimize import OptimizeResult
+from ._constraints import old_bound_to_new, Bounds
+from ._directmodule import direct as _direct  # type: ignore
+
+if TYPE_CHECKING:
+    import numpy.typing as npt
+    from _typeshed import NoneType
+
+__all__ = ['direct']
+
+ERROR_MESSAGES = (
+    "Number of function evaluations done is larger than maxfun={}",
+    "Number of iterations is larger than maxiter={}",
+    "u[i] < l[i] for some i",
+    "maxfun is too large",
+    "Initialization failed",
+    "There was an error in the creation of the sample points",
+    "An error occured while the function was sampled",
+    "Maximum number of levels has been reached.",
+    "Forced stop",
+    "Invalid arguments",
+    "Out of memory",
+)
+
+SUCCESS_MESSAGES = (
+    ("The best function value found is within {} percent "
+     "of the (known) global optimum f_min"),
+    ("The volume of the hyper-rectangle with best function value found "
+     "is below vol_per={}"),
+    ("The measure of the hyper-rectangle with best function value found "
+     "is below sigma_per={}"),
+)
+
+
+def direct(
+    func: Callable[[npt.ArrayLike, Tuple[Any]], float],
+    bounds: Union[Iterable, Bounds],
+    *,
+    args: tuple = (),
+    disp: bool = False,
+    iatol: float = 1e-4,
+    maxfun: int = 20000,
+    maxiter: int = 6000,
+    locally_biased: bool = True,
+    f_min: float = -np.inf,
+    f_min_per: float = 0.01,
+    vol_per: float = -1.0,
+    sigma_per: float = -1.0,
+    callback: Optional[Callable[[npt.ArrayLike], NoneType]] = None
+) -> OptimizeResult:
+    r"""
+    Solve an optimization problem using the DIRECT
+    (Dividing Rectangles) algorithm.
+
+    The algorithm is due to Jones et al. [1]_.
+
+    Parameters
+    ----------
+    func: callable
+        The objective function to be minimized.
+        ``func(x, \*args) -> float``
+        where ``x`` is an 1-D array with shape (n,) and args is a tuple of
+        the fixed parameters needed to completely specify the function.
+    bounds : sequence or `Bounds`
+        Bounds for variables. There are two ways to specify the bounds:
+        1. Instance of `Bounds` class.
+        2. ``(min, max)`` pairs for each element in ``x``, defining the finite
+        lower and upper bounds for the optimizing argument of `func`. It is
+        required to have ``len(bounds) == len(x)``. ``len(bounds)`` is used
+        to determine the number of parameters in ``x``.
+    args : tuple, optional
+        Any additional fixed parameters needed to
+        completely specify the objective function.
+    disp : bool, optional
+        Prints the evaluated `func` at every iteration.
+    iatol : float, optional
+        Ensures sufficient decrease in function value when a new potentially
+        optimal hyperrectangle is chosen. Default is 0.0001.
+    maxfun : int, optional
+        Approximate upper bound on objective function evaluations.
+        Default is 2000.
```

Comment:
The default above is 20,000. 
Also, I have been testing `direct` using SciPy's global optimization benchmarks and find that for all (or almost all) problems, the actual number of function evaluations is a little greater than 20,000. Is this expected?

Reply:
Yes, AFAIK, overshooting the max iterations by a little is normal in DIRECT algorithm. It has two stopping critera (either if it has execeeded max iterations allocated to it or if the value found by it is within some %age of the global minimum passed as input). So, far example if we pass the global as negative infinity then obviously DIRECT will stop only when it has exceeded the max iterations and if we pass global minimum say 0 (if we know that our function is positive always) then it might stop earlier. The reason for its possible slowness as compared to other algorithms is because it treats the function as black box and just uses the outputs from the function only.

Reply:
Would request @dschmitz89 to verify my comment and if possible provide corrections if I made any mistake. Thanks.

Reply:
Good point, Matt, and thanks for running the benchmarks! Looking again at the function signature, DIRECT currently can only run for the maximal number of function evaluations as the other termination criteria are disabled. To stop the optimization, `sigma_per` and `vol_per` probably have to be set. Unfortunately, I cannot find any resources on reasonable defaults. How much work would it be to benchmark a hand ful of parameter sets?

Reply:
I see. I thought `iatol` was a termination criterion, but I guess not?

> How much work and CPU time would it be to benchmark a handful of parameter sets?

Less than a day of work, I think, and the CPU would finish within that same day. For example, the total CPU time for running all benchmarks is less than a minute. (There is some overhead in SciPy's benchmarking system, ASV, though, so it takes a few minutes of wall clock time.)

I can't do that myself right now, but if someone were interested, I have a few thoughts on what I would do.
@swallan @DominicChm does this sound interesting? If not, no reply necessary.

Reply:
I guess all of us had a misunderstanding of `iatol`: looking at [Gablonsky's thesis](https://repository.lib.ncsu.edu/bitstream/handle/1840.16/3920/etd.pdf?sequence=1), it seems that `eps` (which was renamed `iatol` in this PR) "is used to ensure that there is the potential for a sufficient decrease in this hyperrectangle". So it is actually more of an internal hyperparameter which controls DIRECT's iterations. Acc. to the thesis 1e-4 seems to be a good choice.

It actually looks like DIRECT always ran for the maximal number of iterations by default in Jablonsky's implementation as `sigma_per` and `vol_per` were set to -1.

Reply:
I hope to find time to perform these benchmarks over Christmas holidays. 
@andim : did you by any chance perform a benchmark of DIRECT's termination criteria for scipydirect in the past? Literature is basically not available.

Reply:
No, I hadn't benchmarked these! I had just manually tuned the number of iterations to get reasonable results for the specific optimization problem I was working on.